### PR TITLE
Achievements

### PR DESCRIPTION
**Pull request by RylanRCox**
We created a handful of the achievements.

Although, looking at the way our architecture has been laid out and what it's become Yegor and I think that we need to rethink our Achievements. Lots of them are based on semester-long logic. This is tricky because we currently have no way to implement an "end seasonal session" and personally I don't think there is any reason to. It will also cause problems with our all-time scoreboard which would mean that we would have to have lots and lots of duplicate data.

We implemented:
- Gotta Go Fast
- Slowpoke
- Gold
- Silver
- Bronze
- Hat Trick

Even out of these, I think we should drop slowpoke. Anubhav mentions in our first presentation that we should think about our users. I think giving a student a "slowpoke" badge is a little mean-spirited and we should avoid that.

- Rising Star - Awarded to the student who has shown the most improvement from their previous assessment scores.
- Speed Demon - Awarded to the student who consistently submits assessments with the shortest durations.
- Consistency Champion - Given to the student who maintains a high average score over multiple assessments.
- Perfect Score - Given to any student who manages to achieve a perfect score on an assessment.
- Power Streak - Awarded to the student who has the longest streak of getting a top rank in consecutive assessments.
- Unstoppable - Awarded to the student who maintains a top rank for the longest duration without being overtaken.

https://github.com/UBCO-COSC-499-Summer-2023/PrairieLearn-Gamification/pull/321

### CODE REVIEW

**Comment by LouisLP** [comment]
Sure, we can get rid of slowpoke then. Heartbroken might be along the same vein though?
The issue I have with some of these new ones, is that it's hard to quantify. What is "consistently"? What is a "high average"? What is a "top rank"?

Perfect score is a good one. Rising star I like as well (can just make a function that calculates the average of all their scores).

**Comment by RylanRCox** [comment]
The final achievements are as follows.

# Achievements:

1. **Gold** (workspace_premium) - Rank 1 in a live quiz
2. **Silver** (workspace_premium) - Rank 2 in a live quiz
3. **Bronze** (workspace_premium) - Rank 3 in a live quiz
4. **Hat Trick** (school) - Rank 1, three live quizzes in a row
5. **Gotta Go Fast** (rocket_launch) - Fastest finish in a live quiz
6. **Speed Demon** (sprint) - Top 5 fastest times, five live quizzes in a row
7. **Always On Time** (emoji_people) - Attended every live quiz
8. **Unstoppable** (person_celebrate) - Rank 1, five live quizzes in a row

Scrapped Ideas:

- ~~**Powerstreak **~~ (exercise)
  - Top 3 results, five live quizzes in a row.
  - Got scrapped because the cross-checks required are too complex

- ~~**Rising Star**~~ (No Icon)
  - This student had the largest point increase from their last live session to this one.
  - Got scrapped because it seems like too much effort to implement in time constraints.

- ~~**Slowpoke**~~ (elderly)
  - Slowest to submit the quiz
  - Got scrapped because it seemed to mean spirited.

- ~~**Enlightened**~~ (self_improvement)
  - Best score in the semester (only 1 of these is cycled out per semester)
  - Got scrapped because there is no "end of semester" in our system.

- ~~**Heartbroken**~~ (heart_broken)
  - Worst score in the semester (only 1 of these is cycled out per semester)
  - Got scrapped because there is no "end of semester" in our system.

- ~~**The Thinker**~~ (psychology)
  - Highest score with the slowest submission
  - Got scrapped because it seemed too mean spirited.

- ~~**Best of the Best**~~ (hotel_class)
  - Given at the end of the semester to the top scorer
  - Got scrapped because there is no "end of semester in our system.

- ~~**Half-Baked**~~ (cookie)
  - Exactly half the maximum score
  - max_points is rarely defined and hard to set/get

- ~~**Perfect Score**~~ (NEEDS ICON)
  - Perfect score
  -  max_points is rarely defined and hard to set/get

**Comment by RylanRCox** [comment]
As for testing, I built a new set of admin queries that will allow a semester to be simulated with 5 live sessions. Below are the terminal outputs of testing achievements and the tables after running these outputs.

image
This first image demonstrates how the seasonal scoreboard worked, and since the live scoreboard will have too many outputs we just displayed user 11 as a demonstration.

image
This image shows the inputs working on the achievement-related tables. The plr_has_achieved table shows how many achievements a student has gotten. The number of achievements should correspond to the list of accomplishments posted above. All users have 7 because you start with it and lose it when you miss a class. This has been tested and works but is not represented in this image. These scores were generated by running a semester with set values to ensure the same users got first or the fastest to make sure they won the awards.

**Comment by RylanRCox** [comment]
Also completed in this branch was a handful of debugging that was uncovered while testing for achievements.

- The first major error found was that the database's notification sent hundreds of responses per insert overwhelming the client and crashing the database.
  - The solution to this was removing the message within the notification. We're unsure why this was sending duplicate signals but we didn't need the message regardless so we fixed the bug.

- The next major issue was that duplicate issues were being put into the live session credentials table.
  - This was an issue with our insert. Our logic was correct in only inserting with the right session being closed. But the insert itself had no such parameter so it was inserted for all tables and was fixed with a check on is_live.

- The final major issue was that the points column within the seasonal session was being updated with all scores in live session credentials. I.E. each time a user was inserted it would add all their scores together giving a much larger growth
  - The solution to this was similar to the previous in that our logic check was correct but the update statement was not specific enough and was grabbing all live session credential rows with our user's id.

**Comment by RylanRCox** [comment]
Lastly, a new admin query was made in order to test achievements and allows for semester-long testing.

**PLR_1_start_sessions**
- This query runs on click.
- It generates 14 beautiful users having IDs 2-15. This might cause an error if you insert a new user and then run this query.
- It then enrolls each student into course instance 2 which is the test course containing our live assessments.
- It then creates 5 test assessments with IDs ranging from 100-104 to avoid conflicting errors.

**PLR_2_sim_live**
- This requires an assessment ID input which is intended to increment from 100-104 to use the test assessments from PLR_1_start_sessions.
- It creates a live session attached to the assessment input.
- It then gives each student out of the pre-generated students a row with a score. These are inserted into assessment_instances in order to properly test our triggers.

**PLR_3_set_times**
- This requires an assessment ID similar to PLR_2_sim_live which is also intended to increment from 100-104.
- This generated a random time for each user within the live session ranging from 3-12 minutes.
- It also sets user ID 6 to 2 minutes so it is guaranteed to be the fastest. This allows us to check for speed based awards.

**PLR_4_end_live_session**
- This requires an ID for the live session you wish to close. Which typically will be 1-5 if these are run on a fresh container.
- It simply sets the is_live value to false for the live session matching the input ID.

**PLR_2.1_sim_live**
- This query takes the same inputs as PLR_2_sim_live.
- The only difference between this query and it's parent query is that Scott skipped this live session! And because of this he will lose his Always On Time achievement! I.E. This is to test a user losing their Always On Time achievement and can be run instead of any of the other PLR_2_sim_live runs.

### Putting it all together
So putting all this together the steps one should go through to test the DB and achievements are as follows (assuming they have started container and loaded from disk):
1. Run the admin query PLR_1_live_sessions
2. Run the admin query PLR_2_sim_live and set the ID = 100
3. Run the admin query PLR_3_set_times and set the ID = 100
4. Run the admin query PLR_4_end_live_session and set the ID = 1
5. Run the admin query PLR_2_sim_live and set the ID = 101
6. Run the admin query PLR_3_set_times and set the ID = 101
7. Run the admin query PLR_4_end_live_session and set the ID = 2
8. Repeat steps 5 - 7 incrementing the IDs by one ending with the assessment ID = 104 and live session ID = 5.
9.  (Optional) run PLR_2.1_sim_live in place of PLR_2_sim_live to test the Always On Time achievement.

**Comment by LouisLP** [comment]
#31 Caused some conflicts because I had to make changes to `plrStudents` (and the placement of the `server.js` part)

File: apps/prairielearn/src/admin_queries/PLR_1_start_sessions.json
Comment (LouisLP) [inline apps/prairielearn/src/admin_queries/PLR_1_start_sessions.json:1]: Could be a lil more descriptive here

File: apps/prairielearn/src/admin_queries/PLR_1_start_sessions.sql [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      (2, 'fazackernator@ubc.ca', 'Fazackernator'),
+      (3, 'big_ramon@ubc.ca', 'Big Ramon'),
+      (4, 'mouth_man@ubc.ca', 'Mouth of Sauron'),
+      (5, 'YesHoney@ubc.ca', 'Louis Dear'),
+      (6, 'TheYoungGoat@ubc.ca', 'Sheel The Goat'),
+      (7, 'ErenYegor@ubc.ca', 'Babayegor'),
+      (8, 'IAmBecomePL@ubc.ca', 'PL God'),
+      (9, 'LordAndSaviour@ubc.ca', 'The Almighty Nathan'),
+      (10, 'TheReverseFlash@ubc.ca', 'Eobard Thawne'),
+      (11, 'GottaGetThatMoney@ubc.ca', 'Eugene Crabs'),
```
Comment (LouisLP) [inline apps/prairielearn/src/admin_queries/PLR_1_start_sessions.sql]: Isn't it Eugene Krabs with a K?

File: apps/prairielearn/src/admin_queries/PLR_2.1_sim_live.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+    "description": "Simulate the a Live Session, missing Scott",
```
Comment (LouisLP) [inline apps/prairielearn/src/admin_queries/PLR_2.1_sim_live.json]: Spelling mistake, but also I don't understand (from this description) why there's a "start sessions" and then "simulate live session"

File: 20230730085800_PLR_slowpoke_achievement_trigger.sql
Comment (LouisLP) [inline 20230730085800_PLR_slowpoke_achievement_trigger.sql:1]: I thought Slowpoke was supposed to be removed? Also the placement in the folder is strange.

File: apps/prairielearn/src/admin_queries/PLR_2.1_sim_live.json
Comment (LouisLP) [inline apps/prairielearn/src/admin_queries/PLR_2.1_sim_live.json:1]: I'd suggest adding to the description:
> (Optional) run PLR_2.1_sim_live in place of PLR_2_sim_live to test the Always On Time achievement.

File: apps/prairielearn/src/admin_queries/PLR_2_sim_live.json
Comment (LouisLP) [inline apps/prairielearn/src/admin_queries/PLR_2_sim_live.json:1]: Add the descriptions (shortened versions) you wrote in the PR to this JSON:
- This requires an assessment ID input which is intended to increment from 100-104 to use the test assessments from PLR_1_start_sessions.
- It creates a live session attached to the assessment input.
- It then gives each student out of the pre-generated students a row with a score. These are inserted into assessment_instances in order to properly test our triggers.

File: apps/prairielearn/src/migrations/20230730092734_PLR_placement_achievement_triggers.sql
Comment (LouisLP) [inline apps/prairielearn/src/migrations/20230730092734_PLR_placement_achievement_triggers.sql:1]: Interesting way of doing it; the function calls are crazy cool

File: apps/prairielearn/src/pages/plrStudent/plrStudent.js
Comment (LouisLP) [inline apps/prairielearn/src/pages/plrStudent/plrStudent.js:1]: This is gonna cause merge conflicts, but I can check.

**Review by LouisLP — CHANGES_REQUESTED** [review]
Good call numbering them (so that it's clear what to do first). I'm trusting the triggers work because I can't test them right now.

General notes:
- I'd strongly suggest you take the descriptions you added to the [PR here ](https://github.com/LouisLP/PrairieLearn-Ranked/pull/28#issuecomment-1661142437)and turn them into shortened versions for the JSON description fields.
- The achievement triggers could be called `timestamp_PLR_AT_hat_trick.sql` or something similar, so that they're all grouped together (regardless of the timestamp at the beginning)

There's gonna be some small merge conflicts on `plrStudent` stuff (on `plrStudent.js` and `server.js`) but I can check that out.

**Comment by LouisLP** [comment]
@RylanRCox To resolve the merge conflicts, I accepted the current changes (what's already on `master`), because yours seemed to just be some formatting stuff, and the removal of some declaration lines we needed under `plrStudent.js`.

**Review by LouisLP — APPROVED** [review]
LGTM
okay

**Pull request merged**